crates/messages/proto/messages.proto

```diff
@@ -31,137 +19,149 @@ message WorkerState {
+
+  string version = 2;  // "1.0.0"
   optional uint64 stored_bytes = 3;
-  repeated DatasetRanges stored_ranges = 4;
-  bytes signature = 5;
+  string assignment_id = 6;  // "20241008T141245_242da92f7d6c"
+  // Points to the chunks in this worker's assignment list referenced by `assignment_id`.
+  BitString missing_chunks = 7;
 }
 
 message HttpHeader {
+  option deprecated = true;
+
   string name = 1;
   string value = 2;
 }
 
 message AssignedChunk {
+  option deprecated = true;
+
   string path = 1;  // "0000000000/0000808640-0000816499-b0486318"
   repeated uint32 filenames = 2;  // index in known_filenames array
 }
 
 message DatasetChunks {
+  option deprecated = true;
+
   string dataset_id = 1;  // "s3://moonbeam-evm-1"
   string download_url = 3;  // "https://moonbeam-evm-1.sqd-datasets.io/"
   repeated AssignedChunk chunks = 4;
 }
 
 message WorkerAssignment {
+  option deprecated = true;
+
   repeated DatasetChunks dataset_chunks = 1;
   repeated HttpHeader http_headers = 2;
   repeated string known_filenames = 3;  // "blocks.parquet"
 }
 
+// Scheduler -> Worker
 message Pong {
+  option deprecated = true;
+
   bytes ping_hash = 1;
   oneof status {
-    //    google.protobuf.Empty not_registered = 2;
     google.protobuf.Empty unsupported_version = 3;
     string jailed = 6;
     WorkerAssignment active = 7;
   }
 }
 
+// Portal -> Worker
 message Query {// Optional fields enforce serializing default values
-  optional string query_id = 1;
-  optional string dataset = 2;
-  optional string query = 3;
-  optional bool profiling = 4;
-  optional string client_state_json = 5;
-  bytes signature = 6;
+  string query_id = 1;
+  string dataset = 2;  // "czM6Ly9ldGhlcmV1bS1tYWlubmV0"
+  string query = 3;  // 512 KB max
   // If present, these values should be used instead of from_block and to_block in the query contents
   optional Range block_range = 7;
+  uint64 timestamp_ms = 8;
+  bytes signature = 9;  // covers all the above fields + worker's peer ID
 }
 
+// Worker -> Portal
 message QueryResult {
   string query_id = 1;
   oneof result {
-    OkResult ok = 2;
-    string bad_request = 3;
-    string server_error = 4;
-    google.protobuf.Empty no_allocation = 5;
-    google.protobuf.Empty timeout_v1 = 6 [deprecated = true];
-    string timeout = 7;
+    QueryOk ok = 2;
+    QueryError err = 8;
   }
+  optional uint32 retry_after_ms = 9;
 }
 
-message OkResult {
+message QueryOk {
   bytes data = 1;
-  optional bytes exec_plan = 2;
   optional uint64 last_block = 3;
+  QueryResultSummary summary = 4;
 }
 
-message QuerySubmitted {
-  string client_id = 1;
-  string worker_id = 2;
-  string query_id = 3;
-
-  string dataset = 4;
-  string query = 5;
-  bytes query_hash = 6;
+message QueryError {
+  oneof err {
+    string bad_request = 1;
+    string not_found = 2;
+    string server_error = 3;
+    google.protobuf.Empty too_many_requests = 4;
+    google.protobuf.Empty server_overloaded = 5;
+    string timeout = 6;
```

Comment:
Why not? I think every request has a timeout on the transport level

Reply:
Think about that from the client prospective. Can it receive it? How is it different from the generic `server_error`?

Reply:
Yes, it can receive both client and server timeouts. There is always a timeout on the server side — if the query has not been processed in 20s, the request gets interrupted with this error.\
Yes, it should probably be just a server error.